### PR TITLE
Fix admins not able to get files due to permissions

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Allow sysadmins to get files regardless of permission checks.

#### Ticket Link
NONE

#### Release Note
```release-note
Fix issue where sysadmins may not have access to certain files
```

### CODE REVIEW

File: server/channels/api4/file.go [hunk 1]
```diff
@@ -490,14 +490,14 @@ func getFile(c *Context, w http.ResponseWriter, r *http.Request) {
...
 	perm := c.App.SessionHasPermissionToReadChannel(c.AppContext, *c.AppContext.Session(), channel)
-	if info.CreatorId == model.BookmarkFileOwner {
-		if !perm {
+	if !c.IsSystemAdmin() && !perm {
+		if info.CreatorId == model.BookmarkFileOwner {
+			c.SetPermissionError(model.PermissionReadChannelContent)
+			return
+		} else if info.CreatorId != c.AppContext.Session().UserId {
 			c.SetPermissionError(model.PermissionReadChannelContent)
 			return
```
Comment (cpoile) [inline server/channels/api4/file.go:499]: This feels overly complex, and I'm not clear on what if `info.CreatorId == model.BookmarkFileOwner` means in this context -- why should it matter to the current user's permissions if the creator is the owner?

If the rule is that sysadmins should always have permissions, let's roll that into the perm value. 
We have these conditions, (correct me if I'm wrong):
- is sysAdmin: always has permissions
- is creator of the bookmark: has permissions even if they don't have channel permissions -- effectively always has permissions, right?
- has channel read permissions: has permissions

so, `perm = isSysadmin || isCreator || hasChannelPermissions`
right?

**Review by cpoile — CHANGES_REQUESTED** [review]

File: server/channels/api4/file.go (hunk 1 above)
Comment (larkox) [reply server/channels/api4/file.go:499]: I am not familiar myself with bookmarks myself, so I cannot comment on that one (it also feels obscure to me).

I just tried to factor in the "IsSystemAdmin" with just a bit of refactoring.

But I like your idea, let me take a stab at it.

**Comment by marianunez** [comment]
Adding @sbishel and @enahum for a second set of eyes on permissions and bookmark checks

**Comment by sbishel** [comment]
In what case does a `SystemAdmin` fail this check. That is the real issue.
```
	perm := c.App.SessionHasPermissionToReadChannel(c.AppContext, *c.AppContext.Session(), channel)
```
We should not be checking roles, only permissions. I know there are other places we do it, but we shouldn't.

I know `model.BookmarkFileOwner = "Bookmark"` signifies the file is owned by a bookmark. But I'm not sure why it is checked separately.

**Comment by enahum** [comment]
> In what case does a `SystemAdmin` fail this check? That is the real issue.
> 
> ```
> 	perm := c.App.SessionHasPermissionToReadChannel(c.AppContext, *c.AppContext.Session(), channel)
> ```
> 
> We should not be checking roles, only permissions. I know there are other places we do it, but we shouldn't.
> 
> I know `model.BookmarkFileOwner = "Bookmark"` signifies the file is owned by a bookmark (vs a post). But I'm not sure why it is checked separately.

@larkox  @cpoile @sbishel the `model.BookmarkFileOwner = "Bookmark"` is basically a hack we needed to use (that we kinda borrowed from boards) what does it do?

1. specifies that this is a file that was included in a bookmark and is not part of a post
2. prevents the file and file info from being deleted when the job for data retention runs as bookmarks are not part of the data retention process.
3. ensures that a file that was intended for a bookmark is not used for any other purpose

In short the creator is always a constant value and does not really matches any userId

**Comment by larkox** [comment]
Closing in favor of https://github.com/mattermost/mattermost/pull/29656
